### ISSUE TITLE
Inconsistent handling of edge-traversal between routing algorithms

### ISSUE THREAD

**Issue by danpat** [valhalla/valhalla#2885]
Per https://github.com/valhalla/valhalla/pull/2875#discussion_r580720480, we have some divergence in the handling of dead-ends between different algorithms.  This is just a small example of a bigger problem.

We have the following code that all implements slight variants of the graph search:

  - multimodal
  - bidirectional_astar
  - timedep_forward
  - timedep_reverse
  - astar_bss
  - loki/reach
  - centroid
  - isochrone
  - timedistancematrix
  - timedistancematrixbss
  - others??? (map matching?)

Each of these implement *largely* similar code.  When adding things like [fixing uturns at deadends](https://github.com/valhalla/valhalla/pull/2875), [starting routes from closed points](https://github.com/valhalla/valhalla/pull/2846), and others, one ends up needing to duplicate code in multiple places.

We're seeing quite a bit of drift - folks are only making changes to the algorithm that affects the feature they're working on, without realizing (or perhaps being daunted by) how many other places their change needs to be made.

Each of the routing algorithms above should be reviewed.  Common logic needs to be moved into a common base class or template, and algorithm-specific code needs to be kept separate.

**Comment by kevinkreiser** [comment]
@danpat merging as much base functionality into a base class will be a gargantuan refactor considering:

1. use or disuse of hierarchy culling
2. using the correct heuristics depending on uni vs bidirectional
3. different types of edge labels
4. different stopping criteria

I think the end goal is excellent but the journey may be quite arduous :smile: As you noted in #2967 regression tests will be key.

I also wonder if we can somehow move a function at a time into path algorithm? I've tried this a bit in the past (i guess you were alluding to that) and i've found that an approach that worked best was to edit the methods in place (without moving them) until they were nearly identical and everythign was still working, and then finally moving them as the last step. ive also found that there are a decent amount of small differences among implementations which makes this work doubly annoying because the small differences are often edge cases which may not be adequately covered by unit tests.

anyway all that to say i hope we can get some serious consolidation over time :+1:

**Comment by danpat** [comment]
@kevinkreiser yep, I've been staring at the code for a couple of days to try and get enough of a picture to start to take some steps.

I've been doing exactly what you suggested, starting with the isochrone plugin - functions like `ExpandForward` and `ExpandReverse` are relatively straightforward to templatize and condense.

There are inconsistencies of behaviour, and unfixed bugs all over the place - it is going to be difficult to tease out the "someone forgot to fix bug X in implementation Y" from the "implementation Y needs to be a little different because of Z", but we're using Valhalla in production, and there are some real-world downstream effects cropping up from the inconsistent behaviour between services.  So I'm starting the grind :-P

Keep an eye out for PRs.  I'm going to do things in little pieces as I'm not entirely sure what the final destination is going to look like, and it will be so much code moving that I'd rather not drop a huge refactor bomb on the project while lots of other work continues on.

**Comment by kevinkreiser** [comment]
@danpat sounds great to me! ill be on the lookout!

**Comment by nilsnolde** [comment]
At least we managed to simplify the TimeDistanceMatrix a bit more. IMO all the dijkstras need a good look (apart from the whole consistency among _all_ algos), at least the TimeDistanceMatrix could also be based on Dijkstras. I guess the bidirectional & unidirectional variants are very hard to consolidate, for both astar & dijkstra.
